quarkus.jdt.ext/com.redhat.microprofile.jdt.quarkus/src/main/java/com/redhat/microprofile/jdt/internal/quarkus/scheduler/SchedulerUtils.java

```diff
@@ -0,0 +1,128 @@
+*******************************************************************************/
+package com.redhat.microprofile.jdt.internal.quarkus.scheduler;
+
+import java.time.Duration;
+import java.time.format.DateTimeParseException;
+import java.util.regex.Pattern;
+
+/**
+ * Utility class for quarkus scheduler
+ *
+ */
+public class SchedulerUtils {
+
+	// Patterns for each cron part of the cron member expression in order of
+	// SchedulerErrorCode enum ordinal
+	// Note: DAY_OF_MONTH and DAY_OF_WEEK are mutually exclusive
+
+	private static final String DAY_OF_MONTH_ABBREV = "^MON$|^TUE$|^WED$|^THU$|^FRI$|^SAT$|^SUN$";
+	private static final String MONTH_ABBREV = "^JAN$|^FEB$|^MA[RY]$|^APR$|^JU[LN]$|^AUG$|^SEP$|^OCT$|^NOV$|^DEC$";
+	private static final Pattern[] CRON_PATTERNS = {
+			Pattern.compile("(^[1-5]?[0-9]$|\\*|\\?|^\\*\\/\\d+)"), // SECONDS
+			Pattern.compile("(^[1-5]?[0-9]$|\\*|\\?|^\\*\\/\\d+)"), // MINUTES
+			Pattern.compile("(^(2[0-3]|1[0-9]|[1-9])$|\\*|\\?|^\\*\\/\\d+)"), // HOUR
+			Pattern.compile("(^(3[01]|[12][0-9]|[1-9])$|\\*|\\?|^\\*\\/\\d+)"), // DAY_OF_MONTH
+			Pattern.compile(String.format("(%s|^(1[0-2]|[1-9])$|\\*|\\?|^\\*\\/\\d+)", MONTH_ABBREV),
+					Pattern.CASE_INSENSITIVE), // MONTH
+			Pattern.compile(String.format("(%s|[1-7]|\\*|\\?|^\\*\\/\\d+)", DAY_OF_MONTH_ABBREV),
+					Pattern.CASE_INSENSITIVE), // DAY_OF_WEEK
+			Pattern.compile("((^19[7-9][0-9]$|^20[0-9][0-9]$)|\\*|\\?|^\\*\\/\\d+)") // YEAR
+	};
+
+	private static final Pattern TIME_PERIOD_PATTERN = Pattern.compile("\\d+[smhSMH]$");
+
+	private static final Pattern ENV_PATTERN = Pattern.compile("^\\$?\\{[0-9a-zA-Z\\.]*\\}$");
+
+	private static final Pattern LOOSE_ENV_PATTERN = Pattern.compile("^.*\\$?\\{.*\\}.*$");
+
+	private SchedulerUtils() {
+	}
+
+	/**
+	 * Validate the @Scheduled cron member with each cron string part and return an
+	 * error message if necessary
+	 *
+	 * @param cronString the cron member value
+	 * @return the error fault for the cron string validation and null if valid
+	 */
+	public static SchedulerErrorCodes validateCronPattern(String cronString) {
+
+		String[] cronParts = cronString.split("\\s+");
+
+		if (cronParts.length < 6 || cronParts.length > 7) {
+			return SchedulerErrorCodes.INVALID_CRON_LENGTH;
+		}
+
+		SchedulerErrorCodes[] errorCodes = SchedulerErrorCodes.values();
+		for (int i = 0; i < cronParts.length; i++) {
+			if (!CRON_PATTERNS[i].matcher(cronParts[i]).matches()) {
+				return errorCodes[i];
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Validate the string from the @Scheduled member can be parsed to a Duration
+	 * unit
+	 *
+	 * @param timePattern the member value for the time pattern
+	 * @return the INVALID_DURATION_PARSE_PATTERN error code if invalid and null if
+	 *         valid
+	 */
+//	\\d+[smhSMH]
+	public static SchedulerErrorCodes validateDurationParse(String timePattern) {
+		if (!TIME_PERIOD_PATTERN.matcher(timePattern).matches()) {
+			try {
+				Duration.parse(timePattern);
+				return null;
```

Comment:
You ca remove the extra `return null;` on this line. The one at the bottom will be executed if the line above succeeds.